### ISSUE TITLE
grid_sample generalization

### ISSUE THREAD

**Issue by MashiPe** [pytorch/pytorch#100526]
### 🚀 The feature, motivation and pitch

The way that the current implementation of the `grid_sample()` has some downsides. For instance, it enforces the use of a grid with same dimentions as the image. While this is fine for most cases, however there are a few scenarios where we can lose information, for example, lets say that we apply an homography to an image that includes some translation, scaling and rotation, if the pixels of the image ends beyond the [-1,1] boundaries (check image for clarity), we can loss parts of the image. Also, if we want to apply the transformation on a smaller portion of the image (like in the perspective transformation in the [OpenCV Geometric Transformation example](https://docs.opencv.org/4.x/da/d6e/tutorial_py_geometric_transformations.html)) , yes, we could crop the image and then apply the transformation, is not that big of a deal but it's inconvenient and leads to the same loss of information problem. 

![Pytorch Proposal - Cat Example](https://user-images.githubusercontent.com/101341198/235837298-0cf0f194-5b5f-4bac-b1d5-e90ecf3e5316.jpg)

Therefore, my proporsal it's a more generalized version of the `grid_sample()` but offering more flexibility by allowing the use of a grid with different dimentions and provide some kind of position as reference when the transformation is applied on a smaller portion of the image. I believe this is a simple but involved change, so maybe an [RFC](https://github.com/pytorch/rfcs) is needed for further discussion. I would be happy to contribute in the case that the idea is approved.

### Alternatives

You can make an implementation of this idea with numpy and it's fairly easy, however the performance is not that good, also there are extra steps involved in the case that you are working with the gpu are just inconvinient. 

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by drisspg** [comment]
Thank you for opening the issue! It seems like the changes to include this would be fairly substantial.  
I am not as familiar with grid_sample but would it be possible to achieve the same outcome in your example by modyifying the input image, something akine to zero_padding the boundaries? 


> You can make an implementation of this idea with numpy and it's fairly easy, however the performance is not that good, also there are extra steps involved in the case that you are working with the gpu are just inconvinient.

Have you tried writing the decomposed function and and decorating it with [torch.compile](https://pytorch.org/docs/stable/generated/torch.compile.html)?

**Comment by MashiPe** [comment]
Thanks for the reply. Sadly adding padding to the image doesn't give the same result. Here is an example, we take a simple white square and apply a transformation, i'll stay with the homography. So far so good, we get the expected result. Now, if we add the padding, and a green square just for reference, when we apply the transformation we can see that the homography in the outer square is fine, but the white square is not the expected and as far I understand there is no simple way to solve this since in the grid sample [source code](https://github.com/pytorch/pytorch/blob/f064c5aa33483061a48994608d890b968ae53fb5/aten/src/THNN/generic/SpatialGridSamplerBilinear.c#LL70C16-L70C16) the denormalization is applied with the dimensions of the grid. 

![Pytorch Proposal - Squares example](https://github.com/pytorch/pytorch/assets/101341198/4db0b58e-2aec-43ad-ace7-05b1e1bd40d9)


With respect to the decomposed function that's something that I haven tried but I'm no familiar with torch.compile. I'll give it a shot but a more flexible `grid_same()` is something that, in my humble opinion, is useful so adding those changes could help more people. I would be happy to help with the changes as much as needed

**Comment by drisspg** [comment]
> With respect to the decomposed function that's something that I haven tried but I'm no familiar with torch.compile. I'll give it a shot 

If you have the implementation in numpy it should be as simple as porting it to pytorch and then decorating the function with torch.compile().

Your images are very helpful, I appreciate the reference. Correct me if I am wrong, it appears though you are applying the same homography for both the unpadded and padded case. I wouldn't expect this to give the same result. But doesn't there exist another homography that would map the inner region to the output of the non padded mapping?

**Comment by MashiPe** [comment]
> If you have the implementation in numpy it should be as simple as porting it to pytorch and then decorating the function with torch.compile().

I'll try that, hopefully the performance is good enough. However I'm more concern with the part where we calculate the new values, right now I've implemented just like [this](https://github.com/pytorch/pytorch/blob/f064c5aa33483061a48994608d890b968ae53fb5/aten/src/THNN/generic/SpatialGridSamplerBilinear.c#L105), that is, taking the values from the original tensor and then calculate the new value. Not sure if there is a more suitable way to do that that takes advantage of the torch.compile().
 
> Your images are very helpful, I appreciate the reference. Correct me if I am wrong, it appears though you are applying the same homography for both the unpadded and padded case. I wouldn't expect this to give the same result. But doesn't there exist another homography that would map the inner region to the output of the non padded mapping?

I don´t think that's useful, it would be redundant to first apply a transformation of a padded image and then map the inner part to the exact same transformation. Instead just apply the transformation directly, or am I missing something?

**Comment by drisspg** [comment]
Do you have any code pointers? The this links seems to point back to PyTorch impl and I am not sure if thats what you meant. 

>I don´t think that's useful, it would be redundant to first apply a transformation of a padded image and then map the inner part to the exact same transformation.

It seems like the original proposal was to remedy the potential loss of image information after applying a transformation T. If one pads the image, it seems like all the image information is still present, but the resulting transformation T is not the same as if it were applied onto the original(unpaded) image that wasn't clipped at the boundary. All I meant by my above comment is that I wouldn't exepct the *same* transformation T to produce the same output on the unpadded and padded image. But does that mean there exists no transformation T* that would produce the same result on the padded image?

**Comment by MashiPe** [comment]
> Do you have any code pointers? The this links seems to point back to PyTorch impl and I am not sure if thats what you meant.

Sorry for the confusion, what I meant was that once I have denormalized and calculate all the corresponding points of the grid, right now I copy and calculate the corresponding values with a loop as in the Pytorch impl.

```
for h in range(ih):
    for w in range(iw):

        nw_val = (image[int(ix_nw[h,w]),int(iy_nw[h,w])] if int(ix_nw[h,w])>=0 and int(ix_nw[h,w])<w_limit and  
                        int(iy_nw[h,w])>=0 and int(iy_nw[h,w])<h_limit else np.zeros((c)))

        ne_val = (image[int(ix_ne[h,w]),int(iy_ne[h,w])] if int(ix_ne[h,w])>=0 and int(ix_ne[h,w])<w_limit and  
                        int(iy_ne[h,w])>=0 and int(iy_ne[h,w])<h_limit else np.zeros((c)))

        sw_val = (image[int(ix_sw[h,w]),int(iy_sw[h,w])] if int(ix_sw[h,w])>=0 and int(ix_sw[h,w])<w_limit and
                        int(iy_sw[h,w])>=0 and int(iy_sw[h,w])<h_limit else np.zeros((c)))

        se_val = (image[int(ix_se[h,w]),int(iy_se[h,w])] if int(ix_se[h,w])>=0 and int(ix_se[h,w])<w_limit and
                        int(iy_se[h,w])>=0 and int(iy_se[h,w])<h_limit else np.zeros((c)))

        out_val = nw_val * nw[h,w] +ne_val*ne[h,w] + sw_val * sw[h,w] + se_val*se[h,w]

        output[h,w] = out_val
```

Firsts of all, sorry for the ugly code. Now, here `ix_nw`,`iy_nw` and the others are numpy arrays that contain the mapped points after denormalization. So basically the code loops over every element on the grid to get the corresponding point and then calculates the new value. This is the part that I'm more concern since all the other operations can be done with numpy and ported to Pytorch easily and I'm not sure if there is a way to simplify this and speed up the process

> It seems like the original proposal was to remedy the potential loss of image information after applying a transformation T. If one pads the image, it seems like all the image information is still present, but the resulting transformation T is not the same as if it were applied onto the original(unpaded) image that wasn't clipped at the boundary. All I meant by my above comment is that I wouldn't exepct the _same_ transformation T to produce the same output on the unpadded and padded image. But does that mean there exists no transformation T* that would produce the same result on the padded image?

Thanks for the clarification, now I see your point. And technically yes, we could have a transformation T* that in the interior yields the result of the original T, at least with the homography maybe I can come up with some equation that calculates the transformation T* from T since all is matrix algebra. However I'm not sure if that's applicable to all transformations, the affine transformation probably but for something like the TPS I don't think so.

**Comment by MashiPe** [comment]
Not letting this idea die yet. It took me some time to prepare a repo with a complete example and a detailed explanation of what the intended changes look like but [here it is](https://github.com/MashiPe/extended_grid_sample). I hope this can still be discussed and as always I would be happy to contribute to the project.

**Comment by lezcano** [comment]
cc @vfdev-5

**Comment by vfdev-5** [comment]
Hey @MashiPe , if I understand correctly you would like to use grid sampler with an option that expands the output image to make it large enough to hold the entire transformed image. This is already can be done with current API similarly we do in torchvision for rotate with expand: https://pytorch.org/vision/stable/generated/torchvision.transforms.functional.rotate.html#torchvision.transforms.functional.rotate

Here is an example how to achieve that using affine grid and grid sampler:
```python
import torch
import torchvision
torchvision.disable_beta_transforms_warning()


from torchvision.transforms.v2.functional._geometry import _get_inverse_affine_matrix, _compute_affine_output_size, _apply_grid_transform

# Although we already have this function in torchvision, I copy paste the code for better understanding what happens
def _affine_grid(
    theta: torch.Tensor,
    w: int,
    h: int,
    ow: int,
    oh: int,
) -> torch.Tensor:
    # https://github.com/pytorch/pytorch/blob/74b65c32be68b15dc7c9e8bb62459efbfbde33d8/aten/src/ATen/native/
    # AffineGridGenerator.cpp#L18
    # Difference with AffineGridGenerator is that:
    # 1) we normalize grid values after applying theta
    # 2) we can normalize by other image size, such that it covers "extend" option like in PIL.Image.rotate
    dtype = theta.dtype
    device = theta.device

    base_grid = torch.empty(1, oh, ow, 3, dtype=dtype, device=device)
    x_grid = torch.linspace((1.0 - ow) * 0.5, (ow - 1.0) * 0.5, steps=ow, device=device)
    base_grid[..., 0].copy_(x_grid)
    y_grid = torch.linspace((1.0 - oh) * 0.5, (oh - 1.0) * 0.5, steps=oh, device=device).unsqueeze_(-1)
    base_grid[..., 1].copy_(y_grid)
    base_grid[..., 2].fill_(1)

    rescaled_theta = theta.transpose(1, 2).div_(torch.tensor([0.5 * w, 0.5 * h], dtype=dtype, device=device))
    output_grid = base_grid.view(1, oh * ow, 3).bmm(rescaled_theta)
    return output_grid.view(1, oh, ow, 2)


image = torchvision.io.read_image("dog.jpg")

import matplotlib.pyplot as plt
%matplotlib inline

angle = 45
height, width = image.shape[-2:]
center = [0.0, 0.0]

# Here we specify a rotation, translation and scale, keeing shear to be zero and transformation point as image center.
matrix = _get_inverse_affine_matrix(center, -angle, [0.34 * height, -0.23 * width], 1.3, [0.0, 0.0])
theta = torch.tensor(matrix, dtype=torch.float32).reshape(1, 2, 3)

# Affine transformed without expand
ow, oh = width, height
grid = _affine_grid(theta.clone(), w=width, h=height, ow=ow, oh=oh)
output = _apply_grid_transform(image[None, ...], grid, "bilinear", 0)
output = output[0]

# Affine transformed with expand
ow, oh = 2 * width, 2 * height  # Here you can also precompute these values by computing canvas transformation from dst to src
grid = _affine_grid(theta.clone(), w=width, h=height, ow=ow, oh=oh)
output2 = _apply_grid_transform(image[None, ...], grid, "bilinear", 0)
output2 = output2[0]

plt.figure(figsize=(15, 8))
plt.subplot(131)
plt.title("original image")
plt.imshow(image.permute(1, 2, 0))
plt.subplot(132)
plt.title("Affine transformed without expand")
plt.imshow(output.permute(1, 2, 0))
plt.subplot(133)
plt.title("Affine transformed with expand")
plt.imshow(output2.permute(1, 2, 0))
```

![image](https://github.com/pytorch/pytorch/assets/2459423/d373da1a-95c7-4cca-8632-b99bd7110df2)

You can see that 3rd image output size is twice of the original one.

Let me know if it solves your problem.

**Comment by MashiPe** [comment]
@vfdev-5 thanks for the reply. I wasn't aware of that feature and actually is what I'm trying to do. But is it possible to do the same with an Homography or a TPS transformation?

**Comment by vfdev-5** [comment]
@MashiPe as soon as we can construct the grid according to a homography or TPS we can renormalize it by the input image size. Here is, for example, torchvision's implementation to build perspective transformation grid:
https://github.com/pytorch/vision/blob/c9ac3a5b03731fa17d3934b552f308791314602b/torchvision/transforms/v2/functional/_geometry.py#L1305-L1332

It can be improved to take into account expand option and re-centering:
```python
from typing import List


def _perspective_grid(
    coeffs: List[float], 
    w: int, h: int, 
    ow: int, oh: int, 
    dtype: torch.dtype, 
    device: torch.device,
    center = None,
) -> torch.Tensor:
    # https://github.com/python-pillow/Pillow/blob/4634eafe3c695a014267eefdce830b4a825beed7/
    # src/libImaging/Geometry.c#L394

    #
    # x_out = (coeffs[0] * x + coeffs[1] * y + coeffs[2]) / (coeffs[6] * x + coeffs[7] * y + 1)
    # y_out = (coeffs[3] * x + coeffs[4] * y + coeffs[5]) / (coeffs[6] * x + coeffs[7] * y + 1)
    #
    theta1 = torch.tensor(
        [[[coeffs[0], coeffs[1], coeffs[2]], [coeffs[3], coeffs[4], coeffs[5]]]], dtype=dtype, device=device
    )
    theta2 = torch.tensor([[[coeffs[6], coeffs[7], 1.0], [coeffs[6], coeffs[7], 1.0]]], dtype=dtype, device=device)

    d = 0.5
    base_grid = torch.empty(1, oh, ow, 3, dtype=dtype, device=device)
    x_grid = torch.linspace(d, ow + d - 1.0, steps=ow, device=device, dtype=dtype)
    base_grid[..., 0].copy_(x_grid)
    y_grid = torch.linspace(d, oh + d - 1.0, steps=oh, device=device, dtype=dtype).unsqueeze_(-1)
    base_grid[..., 1].copy_(y_grid)
    base_grid[..., 2].fill_(1)

    rescaled_theta1 = theta1.transpose(1, 2).div_(torch.tensor([0.5 * w, 0.5 * h], dtype=dtype, device=device))
    shape = (1, oh * ow, 3)
    output_grid1 = base_grid.view(shape).bmm(rescaled_theta1)
    output_grid2 = base_grid.view(shape).bmm(theta2.transpose(1, 2))

    if center is not None:
        center = torch.tensor(center, dtype=dtype, device=device)
    else:
        center = 1.0

    output_grid = output_grid1.div_(output_grid2).sub_(center)
    return output_grid.view(1, oh, ow, 2)
```

```python
from torchvision.transforms.v2.functional._geometry import _get_perspective_coeffs

height, width = image.shape[-2:]

# startpoints (list of list of ints): List containing four lists of two integers corresponding to four corners
#     ``[top-left, top-right, bottom-right, bottom-left]`` of the original image.
# endpoints (list of list of ints): List containing four lists of two integers corresponding to four corners
#     ``[top-left, top-right, bottom-right, bottom-left]`` of the transformed image.

coeffs = _get_perspective_coeffs(
    [[0, 0], [0, width - 1], [height - 1, width - 1], [height - 1, 0]],
    [[-120, -110], [-30, width + 90], [height - 15, width + 120], [height + 10, 1]]
)

ow, oh = width, height
grid = _perspective_grid(coeffs, w=width, h=height, ow=ow, oh=oh, dtype=torch.float32, device="cpu")
output = _apply_grid_transform(image[None, ...], grid, "bilinear", 0)
output = output[0]

ow, oh = 2 * width, 2 * height
grid = _perspective_grid(
    coeffs, w=width, h=height, ow=ow, oh=oh, 
    dtype=torch.float32, device="cpu",
    center = [1.5, 1.5]
)
output2 = _apply_grid_transform(image[None, ...], grid, "bilinear", 0)
output2 = output2[0]
```

![image](https://github.com/pytorch/pytorch/assets/2459423/36d18619-cd07-4750-ba88-a81f59ac0081)

**Comment by MashiPe** [comment]
Looks like I need to improve my google skills because I never found that `_perspective_grid` function when I was looking for a better way to do this and for what I can see It's fairly easy to implement other geometric transformations if needed. Thank you so much for your time and effort.

**Comment by vfdev-5** [comment]
Glad it works for you. Let me close this issue as solved. If anything is still unclear about this topic please open another issue.
